src/SessionManager/ProgramSettings.cc

```diff
@@ -34,6 +38,141 @@ void applyOptionalArgument(T& val, const string& argument_name, const cxxopts::P
 }
 
 ProgramSettings::ProgramSettings(int argc, char** argv) {
+    if (argc > 1) {
+        spdlog::info("Using command-line settings");
+    }
+    ApplyCommandLineSettings(argc, argv);
+
+    const fs::path user_settings_path = fs::path(getenv("HOME")) / CARTA_USER_FOLDER_PREFIX / "backend.json";
+    const fs::path system_settings_path = "/etc/carta/backend.json";
+
+    json settings;
+    if (fs::exists(system_settings_path) && !no_system_config) {
+        settings = JSONConfigSettings(system_settings_path.string());
+        system_settings_json_exists = true;
+        spdlog::info("System settings found. Command-line arguments and will override system settings");
+        spdlog::info("If you want to ignore system settings, use --no_system_config as command-line argument.");
+    }
+
+    if (fs::exists(user_settings_path) && !no_user_config) {
+        auto user_settings = JSONConfigSettings(user_settings_path.string());
+        user_settings_json_exists = true;
+        spdlog::info(
+            "User settings found, applying on top of system settings. Note that Command-line arguments will override system and user "
+            "settings");
+        spdlog::info("If you want to ignore system settings, use --no_user_config as command-line argument.");
+        settings.merge_patch(user_settings); // user on top of system
+    }
+
+    if (system_settings_json_exists || user_settings_json_exists) {
+        settings.merge_patch(command_line_settings); // force command-line on top of user and sytem
+        SetSettingsFromJSON(settings);
+    }
+}
+
+json ProgramSettings::JSONConfigSettings(const std::string& json_file_path) {
+    std::ifstream ifs(json_file_path, std::ifstream::in);
+    json j;
+    try {
+        j = json::parse(ifs);
+    } catch (json::exception err) {
+        spdlog::warn("Config file {} has problems, please check", json_file_path);
+        spdlog::warn(err.what());
+    }
+    return j;
+}
+
+void ProgramSettings::SetSettingsFromJSON(const json& j) {
+    ValidateJSON(j); // check for invalid keys, store them in std::set for easy counting
+    // logic here is
+    // value = check invalid, if invalid then default value, otherwise check json and assign its value, otherwise default
+    if (!invalid_keys.count("verbosity")) {
+        verbosity = j.value("verbosity", verbosity);
+    }
```

Comment:
As I recall we changed this from a slightly different syntax to be less verbose, but these invalid checks reintroduce the same verbosity.

I'm also not sure that `ValidateJSON` should be a separate function to this and use `invalid keys` to store intermediate state. Ideally validation and setting the value should be done once in the same loop(s).

Reply:
We changed the code to another place to reduce verbosity. It was possible because the verbosity overhead came from string values but not from variables with different types. I think this validation check is important, and it's good to have it in a separate method. Otherwise, we would need to add another library to check against the JSON schema. If we mix the validation part into the settings assignment, I think the verbosity will only increase.

Reply:
We changed to `.value` to avoid repeated `.contains` checks, but we have now gone back to the same problem but with a different membership check. I'll elaborate on my suggestion in my other comment.

Reply:
I do remember that, but this adds another level of check for some type of safety. So, it's in principle a new change.

Reply:
Yes, I understand -- I'm not objecting to the check, but to the verbosity of the explicit repetition (which is the kind of verbosity the original change removed). Hence my suggestion to combine this and the validation function into a single pass which uses maps of keys to attributes.